File: zig/entity/moon.zig
// Moon entity client (generated — mirrors the go/rust Entity fragment).

const std = @import("std");
const vs = @import("voxgig-struct");
const h = @import("../core/helpers.zig");
const errmod = @import("../core/error.zig");
const types = @import("../core/types.zig");
const ctxmod = @import("../core/context.zig");
const utility_mod = @import("../core/utility.zig");
const sdk = @import("../core/sdk.zig");

const Value = h.Value;
const Context = ctxmod.Context;
const CtxSpec = ctxmod.CtxSpec;
const Utility = utility_mod.Utility;
const OpResult = types.OpResult;
const OutVal = types.OutVal;
const Entity = types.Entity;

// Every operation resolves to the ENTITY, not the raw data — `list` to a
// slice of them, one per record; the record is reached through `data()`.
// See AGENTS.md "Entity operations return ENTITIES".
//
// `Value` cannot carry an entity (it is a closed data union) and the shared
// `OpResult` cannot name a per-entity type, so each entity declares its own
// result unions and the CONTRACT lives in those signatures.
pub const EntResult = union(enum) {
    ok: *MoonEntity,
    err: *errmod.SolardemoError,
};

pub const EntListResult = union(enum) {
    ok: []*MoonEntity,
    err: *errmod.SolardemoError,
};

pub const MoonEntity = struct {
    name: []const u8 = "moon",
    client: *sdk.SolardemoSDK,
    utility: *Utility,
    entopts: Value,
    data: Value,
    mtch: Value,
    entctx: ?*Context = null,
    // Set once a successful `remove` resolves on this instance.
    deleted: bool = false,

    pub fn new(client: *sdk.SolardemoSDK, entopts_in: Value) *MoonEntity {
        const entopts: Value = switch (entopts_in) {
            .object => entopts_in,
            else => h.omap(),
        };
        const act = h.get_bool(entopts, "active");
        if (act == null) {
            h.setp(entopts, "active", h.vbool(true));
        } else if (act.? != false) {
            h.setp(entopts, "active", h.vbool(true));
        }

        const e = h.A().create(MoonEntity) catch unreachable;
        e.* = .{
            .name = "moon",
            .client = client,
            .utility = client.get_utility(),
            .entopts = entopts,
            .data = h.omap(),
            .mtch = h.omap(),
            .entctx = null,
        };

        const entctx = e.utility.make_context(CtxSpec{
            .entity = e.asEntity(),
            .entopts = entopts,
        }, client.get_root_ctx());

        e.utility.feature_hook(entctx, "PostConstructEntity");
        e.entctx = entctx;
        return e;
    }

    fn ent_ctx(self: *MoonEntity) *Context {
        return self.entctx orelse unreachable;
    }

    fn opError(self: *MoonEntity, ctx: *Context) OpResult {
        const v = self.utility.make_error(ctx) catch return .{ .err = ctx.pending_err.? };
        return .{ .ok = v };
    }

    fn doneResult(self: *MoonEntity, ctx: *Context) OpResult {
        const v = self.utility.done(ctx) catch return .{ .err = ctx.pending_err.? };
        return .{ .ok = v };
    }

    // Runs the pipeline and hands back THIS entity: run_op has just absorbed
    // the result into it. See AGENTS.md "Entity operations return ENTITIES".
    fn run_op_ent(self: *MoonEntity, ctx: *Context, post_done: *const fn (*MoonEntity, *Context) void) EntResult {
        return switch (self.run_op(ctx, post_done)) {
            .err => |e| EntResult{ .err = e },
            .ok => EntResult{ .ok = self },
        };
    }

    // `remove` resolves to the entity, marked. The instance KEEPS the data it
    // held — a caller can still read what was deleted — but it is no longer a
    // live record.
    pub fn mark_deleted(self: *MoonEntity) void {
        self.deleted = true;
    }

    pub fn is_deleted(self: *MoonEntity) bool {
        return self.deleted;
    }

    fn run_op(self: *MoonEntity, ctx: *Context, post_done: *const fn (*MoonEntity, *Context) void) OpResult {
        const utility = self.utility;

        self.utility.feature_hook(ctx, "PrePoint");
        const point = utility.make_point(ctx) catch return self.opError(ctx);
        ctx.out_set("point", OutVal{ .val = point });

        self.utility.feature_hook(ctx, "PreSpec");
        const spec = utility.make_spec(ctx) catch return self.opError(ctx);
        ctx.out_set("spec", OutVal{ .spec = spec });

        self.utility.feature_hook(ctx, "PreRequest");
        const resp = utility.make_request(ctx) catch return self.opError(ctx);
        ctx.out_set("request", OutVal{ .response = resp });

        self.utility.feature_hook(ctx, "PreResponse");
        const resp2 = utility.make_response(ctx) catch return self.opError(ctx);
        ctx.out_set("response", OutVal{ .response = resp2 });

        self.utility.feature_hook(ctx, "PreResult");
        const result = utility.make_result(ctx) catch return self.opError(ctx);
        ctx.out_set("result", OutVal{ .result = result });

        self.utility.feature_hook(ctx, "PreDone");
        post_done(self, ctx);

        return self.doneResult(ctx);
    }

    /// Streaming operation. Runs `action` through the full pipeline and
    /// returns a slice of the result items, so the `streaming` feature's
    /// incremental output is reachable from a generated entity (a normal op
    /// call materialises the whole result). This runtime is synchronous and
    /// zig has no built-in lazy iterators, so the returned slice is a
    /// materialised cursor the caller walks. `callopts` parameterises the
    /// call: inbound yields the streaming feature's items when active, else
    /// the materialised items; outbound attaches an iterable `body` to the
    /// request (reqdata `body$`); `ctrl` threads pipeline control.
    pub fn stream(self: *MoonEntity, action: []const u8, args: Value, callopts: Value) []Value {
        const utility = self.utility;

        const stream_opts: Value = switch (callopts) {
            .object => callopts,
            else => h.omap(),
        };

        const ctrl: Value = switch (h.to_map(h.getp(stream_opts, "ctrl"))) {
            .object => h.to_map(h.getp(stream_opts, "ctrl")),
            else => h.omap(),
        };
        h.setp(ctrl, "stream", stream_opts);

        const reqmatch: Value = switch (args) {
            .object => args,
            else => h.omap(),
        };

        const ctx = utility.make_context(CtxSpec{
            .opname = action,
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqmatch = reqmatch,
        }, self.ent_ctx());

        // Outbound: attach a caller `body` so the transport can stream a
        // request payload (reqdata `body$`).
        const body = h.getp(stream_opts, "body");
        if (!h.is_noval(body)) {
            const reqdata: Value = switch (ctx.reqdata) {
                .object => ctx.reqdata,
                else => h.omap(),
            };
            h.setp(reqdata, "body$", body);
            ctx.reqdata = reqdata;
        }

        // Run the same pipeline as run_op, firing the feature hooks (the
        // streaming feature attaches result.stream on PreResult).
        utility.feature_hook(ctx, "PrePoint");
        const point = utility.make_point(ctx) catch return &.{};
        ctx.out_set("point", OutVal{ .val = point });

        utility.feature_hook(ctx, "PreSpec");
        const spec = utility.make_spec(ctx) catch return &.{};
        ctx.out_set("spec", OutVal{ .spec = spec });

        utility.feature_hook(ctx, "PreRequest");
        const resp = utility.make_request(ctx) catch return &.{};
        ctx.out_set("request", OutVal{ .response = resp });

        utility.feature_hook(ctx, "PreResponse");
        const resp2 = utility.make_response(ctx) catch return &.{};
        ctx.out_set("response", OutVal{ .response = resp2 });

        utility.feature_hook(ctx, "PreResult");
        const result = utility.make_result(ctx) catch return &.{};
        ctx.out_set("result", OutVal{ .result = result });

        utility.feature_hook(ctx, "PreDone");

        // Inbound: prefer the streaming feature's incremental producer; else
        // fall back to the materialised items so stream always yields.
        if (ctx.result) |res| {
            if (res.stream) |sf| {
                return sf.call(sf.ctx);
            }
        }

        const data = utility.done(ctx) catch return &.{};
        if (data == .array) {
            return data.array.data.items;
        } else if (!h.is_noval(data)) {
            var out = std.ArrayList(Value).init(h.A());
            out.append(data) catch {};
            return out.toOwnedSlice() catch &.{};
        }
        return &.{};
    }

    // ---- Entity interface ----

    fn e_of(p: *anyopaque) *MoonEntity {
        return @ptrCast(@alignCast(p));
    }
    fn v_get_name(p: *anyopaque) []const u8 {
        return e_of(p).name;
    }
    fn v_make(p: *anyopaque) Entity {
        const self = e_of(p);
        const opts = h.omap();
        if (self.entopts == .object) {
            var it = self.entopts.object.iterator();
            while (it.next()) |kv| h.setp(opts, kv.key_ptr.*, kv.value_ptr.*);
        }
        return MoonEntity.new(self.client, opts).asEntity();
    }
    fn v_data(p: *anyopaque, args: ?Value) Value {
        return e_of(p).data_impl(args);
    }
    fn v_matchv(p: *anyopaque, args: ?Value) Value {
        return e_of(p).matchv_impl(args);
    }

    const entity_vtable = Entity.VTable{
        .get_name = v_get_name,
        .make = v_make,
        .data = v_data,
        .matchv = v_matchv,
    };

    pub fn asEntity(self: *MoonEntity) Entity {
        return .{ .ptr = @ptrCast(self), .vtable = &entity_vtable };
    }

    fn data_impl(self: *MoonEntity, args: ?Value) Value {
        if (args) |arg| {
            if (!h.is_noval(arg)) {
                const cloned = h.to_map(h.clone(arg));
                self.data = switch (cloned) {
                    .object => cloned,
                    else => h.omap(),
                };
                self.utility.feature_hook(self.ent_ctx(), "SetData");
            }
        }
        self.utility.feature_hook(self.ent_ctx(), "GetData");
        return h.clone(self.data);
    }

    fn matchv_impl(self: *MoonEntity, args: ?Value) Value {
        if (args) |arg| {
            if (!h.is_noval(arg)) {
                const cloned = h.to_map(h.clone(arg));
                self.mtch = switch (cloned) {
                    .object => cloned,
                    else => h.omap(),
                };
                self.utility.feature_hook(self.ent_ctx(), "SetMatch");
            }
        }
        self.utility.feature_hook(self.ent_ctx(), "GetMatch");
        return h.clone(self.mtch);
    }

    // ---- CRUD operations ----


    pub fn load(self: *MoonEntity, reqmatch: Value, ctrl: Value) EntResult {
        const ctx = self.utility.make_context(CtxSpec{
            .opname = "load",
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqmatch = reqmatch,
        }, self.ent_ctx());
        return self.run_op_ent(ctx, load_post_done);
    }
    
    fn load_post_done(self: *MoonEntity, ctx: *Context) void {
        if (ctx.result) |result| {
            const resmatch = result.resmatch;
            const resdata = result.resdata;
            if (resmatch == .object) self.mtch = resmatch;
            if (!h.is_noval(resdata)) {
                const cm = h.to_map(h.clone(resdata));
                self.data = if (cm == .object) cm else h.omap();
            }
        }
    }
    


    pub fn list(self: *MoonEntity, reqmatch: Value, ctrl: Value) EntListResult {
        const ctx = self.utility.make_context(CtxSpec{
            .opname = "list",
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqmatch = reqmatch,
        }, self.ent_ctx());
        const out = switch (self.run_op(ctx, list_post_done)) {
            .err => |e| return EntListResult{ .err = e },
            .ok => |v| v,
        };
    
        // `list` resolves to one ENTITY per record. make_result cannot build them
        // here — it works in Value, which has no slot for an entity — so the op
        // does, mirroring what the dynamic targets get from make_result.
        var items = std.ArrayList(*MoonEntity).init(h.A());
        if (out == .array) {
            for (out.array.data.items) |entry| {
                const ent = MoonEntity.new(self.client, h.clone(self.entopts));
                if (entry == .object) {
                    _ = ent.data_impl(entry);
                }
                items.append(ent) catch {};
            }
        }
    
        return EntListResult{ .ok = items.toOwnedSlice() catch &[_]*MoonEntity{} };
    }
    
    fn list_post_done(self: *MoonEntity, ctx: *Context) void {
        if (ctx.result) |result| {
            const resmatch = result.resmatch;
            if (resmatch == .object) self.mtch = resmatch;
        }
    }
    


    pub fn create(self: *MoonEntity, reqdata: Value, ctrl: Value) EntResult {
        const ctx = self.utility.make_context(CtxSpec{
            .opname = "create",
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqdata = reqdata,
        }, self.ent_ctx());
        return self.run_op_ent(ctx, create_post_done);
    }
    
    fn create_post_done(self: *MoonEntity, ctx: *Context) void {
        if (ctx.result) |result| {
            const resdata = result.resdata;
            if (!h.is_noval(resdata)) {
                const cm = h.to_map(h.clone(resdata));
                self.data = if (cm == .object) cm else h.omap();
            }
        }
    }
    


    pub fn update(self: *MoonEntity, reqdata: Value, ctrl: Value) EntResult {
        const ctx = self.utility.make_context(CtxSpec{
            .opname = "update",
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqdata = reqdata,
        }, self.ent_ctx());
        return self.run_op_ent(ctx, update_post_done);
    }
    
    fn update_post_done(self: *MoonEntity, ctx: *Context) void {
        if (ctx.result) |result| {
            const resmatch = result.resmatch;
            const resdata = result.resdata;
            if (resmatch == .object) self.mtch = resmatch;
            if (!h.is_noval(resdata)) {
                const cm = h.to_map(h.clone(resdata));
                self.data = if (cm == .object) cm else h.omap();
            }
        }
    }
    


    pub fn remove(self: *MoonEntity, reqmatch: Value, ctrl: Value) EntResult {
        const ctx = self.utility.make_context(CtxSpec{
            .opname = "remove",
            .ctrl = ctrl,
            .mtch = self.mtch,
            .data = self.data,
            .reqmatch = reqmatch,
        }, self.ent_ctx());
        const res = self.run_op_ent(ctx, remove_post_done);
        // A removed entity keeps its data but is no longer a live record.
        if (res == .ok) self.mark_deleted();
        return res;
    }
    
    fn remove_post_done(self: *MoonEntity, ctx: *Context) void {
        if (ctx.result) |result| {
            const resmatch = result.resmatch;
            const resdata = result.resdata;
            if (resmatch == .object) self.mtch = resmatch;
            if (!h.is_noval(resdata)) {
                const cm = h.to_map(h.clone(resdata));
                self.data = if (cm == .object) cm else h.omap();
            }
        }
    }
    
};

File: rust/entity/moon.rs
// Moon entity client (generated — mirrors the go Entity fragment).

// Ops present vary per entity, so some shared imports may go unused.
#![allow(unused_imports)]

use std::cell::RefCell;
use std::rc::Rc;

use crate::core::context::{Context, CtxSpec};
use crate::core::error::SolardemoError;
use crate::core::helpers::{get_bool, getp, setp, to_map};
use crate::core::sdk::SolardemoSDK;
use crate::core::types::{Entity, SolardemoEntity};
use crate::core::utility_type::Utility;
use crate::utility::voxgigstruct as vs;
use crate::utility::voxgigstruct::Value;

pub struct MoonEntity {
    name: String,
    #[allow(dead_code)]
    client: Rc<SolardemoSDK>,
    utility: Rc<Utility>,
    entopts: Value,
    data: RefCell<Value>,
    mtch: RefCell<Value>,
    entctx: RefCell<Option<Rc<Context>>>,
    // Set once a successful `remove` resolves on this instance.
    deleted: RefCell<bool>,
}

impl MoonEntity {
    pub fn new(client: &Rc<SolardemoSDK>, entopts: Value) -> Rc<MoonEntity> {
        let entopts = match entopts {
            Value::Map(m) => Value::Map(m),
            _ => Value::empty_map(),
        };
        if get_bool(&entopts, "active").is_none() {
            setp(&entopts, "active", Value::Bool(true));
        } else if get_bool(&entopts, "active") != Some(false) {
            setp(&entopts, "active", Value::Bool(true));
        }

        let e = Rc::new(MoonEntity {
            name: "moon".to_string(),
            client: client.clone(),
            utility: client.get_utility(),
            entopts: entopts.clone(),
            data: RefCell::new(Value::empty_map()),
            mtch: RefCell::new(Value::empty_map()),
            entctx: RefCell::new(None),
            deleted: RefCell::new(false),
        });

        let entctx = e.utility.make_context(
            CtxSpec {
                entity: Some(e.clone() as Rc<dyn Entity>),
                entopts: Some(entopts),
                ..Default::default()
            },
            Some(&client.get_root_ctx()),
        );

        e.utility.feature_hook(&entctx, "PostConstructEntity");

        *e.entctx.borrow_mut() = Some(entctx);

        e
    }

    fn ent_ctx(&self) -> Rc<Context> {
        self.entctx
            .borrow()
            .clone()
            .expect("entity context not initialised")
    }

    // Runs the pipeline and returns the terminal result VALUE. The entity
    // contract lives one level up, in the op methods below: they call this,
    // then hand back the entity (see AGENTS.md). It is split that way because
    // `Value` is a closed data union that cannot carry an entity.
    fn run_op(
        &self,
        ctx: &Rc<Context>,
        post_done: &dyn Fn(&Rc<Context>),
    ) -> Result<Value, SolardemoError> {
        let utility = &self.utility;

        self.utility.feature_hook(ctx, "PrePoint");
        let point = match utility.make_point(ctx) {
            Ok(p) => p,
            Err(err) => return utility.make_error(ctx, Some(err)),
        };
        ctx.out_set("point", crate::core::types::OutVal::Val(point));

        self.utility.feature_hook(ctx, "PreSpec");
        let spec = match utility.make_spec(ctx) {
            Ok(s) => s,
            Err(err) => return utility.make_error(ctx, Some(err)),
        };
        ctx.out_set("spec", crate::core::types::OutVal::Spec(spec));

        self.utility.feature_hook(ctx, "PreRequest");
        let resp = match utility.make_request(ctx) {
            Ok(r) => r,
            Err(err) => return utility.make_error(ctx, Some(err)),
        };
        ctx.out_set("request", crate::core::types::OutVal::Response(resp));

        self.utility.feature_hook(ctx, "PreResponse");
        let resp2 = match utility.make_response(ctx) {
            Ok(r) => r,
            Err(err) => return utility.make_error(ctx, Some(err)),
        };
        ctx.out_set("response", crate::core::types::OutVal::Response(resp2));

        self.utility.feature_hook(ctx, "PreResult");
        let result = match utility.make_result(ctx) {
            Ok(r) => r,
            Err(err) => return utility.make_error(ctx, Some(err)),
        };
        ctx.out_set("result", crate::core::types::OutVal::Result(result));

        self.utility.feature_hook(ctx, "PreDone");
        post_done(ctx);

        utility.done(ctx)
    }

    /// Streaming operation. Runs `action` through the full pipeline and
    /// returns an iterator over the result items, so the `streaming`
    /// feature's incremental output is reachable from a generated entity (a
    /// normal op call materialises the whole result). This runtime is
    /// synchronous, so the returned iterator is a lazy cursor over items the
    /// pipeline produced. `callopts` parameterises the call:
    ///   - inbound (download): iterate items/chunks from the streaming
    ///     feature when active, else the materialised items;
    ///   - outbound (upload): a `body` in `callopts` is attached to the
    ///     request (reqdata `body$`) so the transport can stream a payload;
    ///   - `ctrl` (pipeline control) threads pipeline options.
    pub fn stream(
        &self,
        action: &str,
        args: Value,
        callopts: Value,
    ) -> Result<std::vec::IntoIter<Value>, SolardemoError> {
        let stream_opts = match &callopts {
            Value::Map(_) => callopts.clone(),
            _ => Value::empty_map(),
        };

        let ctrl = match to_map(&getp(&stream_opts, "ctrl")) {
            Value::Map(m) => Value::Map(m),
            _ => Value::empty_map(),
        };
        setp(&ctrl, "stream", stream_opts.clone());

        // `args` carries the op's request match (list/load); pass it through
        // as reqmatch so the same pipeline the op methods run is exercised.
        let reqmatch = match to_map(&args) {
            Value::Map(m) => Value::Map(m),
            _ => Value::empty_map(),
        };

        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some(action.to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqmatch: Some(reqmatch),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );

        // Outbound: attach a caller `body` so the transport can stream a
        // request payload (reqdata `body$`).
        let body = getp(&stream_opts, "body");
        if !body.is_noval() && !body.is_null() {
            let reqdata = match ctx.reqdata.borrow().clone() {
                Value::Map(m) => Value::Map(m),
                _ => Value::empty_map(),
            };
            setp(&reqdata, "body$", body);
            *ctx.reqdata.borrow_mut() = reqdata;
        }

        // Run the same pipeline as run_op.
        self.utility.feature_hook(&ctx, "PrePoint");
        let point = self.utility.make_point(&ctx)?;
        ctx.out_set("point", crate::core::types::OutVal::Val(point));

        self.utility.feature_hook(&ctx, "PreSpec");
        let spec = self.utility.make_spec(&ctx)?;
        ctx.out_set("spec", crate::core::types::OutVal::Spec(spec));

        self.utility.feature_hook(&ctx, "PreRequest");
        let resp = self.utility.make_request(&ctx)?;
        ctx.out_set("request", crate::core::types::OutVal::Response(resp));

        self.utility.feature_hook(&ctx, "PreResponse");
        let resp2 = self.utility.make_response(&ctx)?;
        ctx.out_set("response", crate::core::types::OutVal::Response(resp2));

        self.utility.feature_hook(&ctx, "PreResult");
        let result = self.utility.make_result(&ctx)?;
        ctx.out_set("result", crate::core::types::OutVal::Result(result));

        self.utility.feature_hook(&ctx, "PreDone");

        // Inbound: prefer the streaming feature's incremental producer; else
        // fall back to the materialised items so `stream` always yields.
        let cur = ctx.result.borrow().clone();
        if let Some(res) = &cur {
            let streamfn = res.borrow().stream.clone();
            if let Some(sf) = streamfn {
                return Ok(sf().into_iter());
            }
        }

        let data = self.utility.done(&ctx)?;
        let items: Vec<Value> = match data {
            Value::List(l) => l.borrow().iter().cloned().collect(),
            Value::Noval | Value::Null => Vec::new(),
            other => vec![other],
        };
        Ok(items.into_iter())
    }
}

impl Entity for MoonEntity {
    fn get_name(&self) -> String {
        self.name.clone()
    }

    // `remove` resolves to the entity, marked. The instance KEEPS the data
    // it held — a caller can still read what was deleted — but it is no
    // longer a live record.
    fn mark_deleted(&self) {
        *self.deleted.borrow_mut() = true;
    }

    fn deleted(&self) -> bool {
        *self.deleted.borrow()
    }

    fn make(&self) -> Rc<dyn Entity> {
        let opts = Value::empty_map();
        if let Value::Map(m) = &self.entopts {
            for (k, v) in m.borrow().iter() {
                setp(&opts, k, v.clone());
            }
        }
        MoonEntity::new(&self.client, opts) as Rc<dyn Entity>
    }

    fn data(&self, args: Option<&Value>) -> Value {
        if let Some(arg) = args {
            if !arg.is_noval() && !arg.is_null() {
                let cloned = to_map(&vs::clone(arg));
                *self.data.borrow_mut() = match cloned {
                    Value::Map(m) => Value::Map(m),
                    _ => Value::empty_map(),
                };
                self.utility.feature_hook(&self.ent_ctx(), "SetData");
            }
        }

        self.utility.feature_hook(&self.ent_ctx(), "GetData");
        vs::clone(&self.data.borrow())
    }

    fn matchv(&self, args: Option<&Value>) -> Value {
        if let Some(arg) = args {
            if !arg.is_noval() && !arg.is_null() {
                let cloned = to_map(&vs::clone(arg));
                *self.mtch.borrow_mut() = match cloned {
                    Value::Map(m) => Value::Map(m),
                    _ => Value::empty_map(),
                };
                self.utility.feature_hook(&self.ent_ctx(), "SetMatch");
            }
        }

        self.utility.feature_hook(&self.ent_ctx(), "GetMatch");
        vs::clone(&self.mtch.borrow())
    }
}

impl SolardemoEntity for MoonEntity {

    fn load(self: &Rc<Self>, reqmatch: Value, ctrl: Value) -> Result<Rc<Self>, SolardemoError> {
        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some("load".to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqmatch: Some(reqmatch),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );
    
        self.run_op(&ctx, &|ctx| {
            if let Some(result) = ctx.result.borrow().clone() {
                let (resmatch, resdata) = {
                    let r = result.borrow();
                    (r.resmatch.clone(), r.resdata.clone())
                };
                if let Value::Map(_) = resmatch {
                    *self.mtch.borrow_mut() = resmatch;
                }
                if !resdata.is_noval() && !resdata.is_null() {
                    *self.data.borrow_mut() = match to_map(&vs::clone(&resdata)) {
                        Value::Map(m) => Value::Map(m),
                        _ => Value::empty_map(),
                    };
                }
            }
        })?;
    
        // The operation resolves to THIS entity: `run_op` has just absorbed the
        // result into it, and the caller reaches the record through `.data(None)`.
        // See AGENTS.md "Entity operations return ENTITIES".
    
        Ok(self.clone())
    }
    


    fn list(self: &Rc<Self>, reqmatch: Value, ctrl: Value) -> Result<Vec<Rc<Self>>, SolardemoError> {
        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some("list".to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqmatch: Some(reqmatch),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );
    
        let out = self.run_op(&ctx, &|ctx| {
            if let Some(result) = ctx.result.borrow().clone() {
                let resmatch = result.borrow().resmatch.clone();
                if let Value::Map(_) = resmatch {
                    *self.mtch.borrow_mut() = resmatch;
                }
            }
        })?;
    
        // `list` resolves to one ENTITY per record. makeResult cannot build them
        // here — it works in `Value`, which is a closed data union with no slot
        // for an entity — so the op does, mirroring what the dynamic targets get
        // from makeResult. See AGENTS.md "Entity operations return ENTITIES".
        let mut items: Vec<Rc<Self>> = Vec::new();
        if let Value::List(list) = &out {
            for entry in list.borrow().iter() {
                let ent = MoonEntity::new(&self.client, vs::clone(&self.entopts));
                if let Value::Map(_) = entry {
                    ent.data(Some(entry));
                }
                items.push(ent);
            }
        }
    
        Ok(items)
    }
    


    fn create(self: &Rc<Self>, reqdata: Value, ctrl: Value) -> Result<Rc<Self>, SolardemoError> {
        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some("create".to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqdata: Some(reqdata),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );
    
        self.run_op(&ctx, &|ctx| {
            if let Some(result) = ctx.result.borrow().clone() {
                let resdata = result.borrow().resdata.clone();
                if !resdata.is_noval() && !resdata.is_null() {
                    *self.data.borrow_mut() = match to_map(&vs::clone(&resdata)) {
                        Value::Map(m) => Value::Map(m),
                        _ => Value::empty_map(),
                    };
                }
            }
        })?;
    
        // The operation resolves to THIS entity: `run_op` has just absorbed the
        // result into it, and the caller reaches the record through `.data(None)`.
        // See AGENTS.md "Entity operations return ENTITIES".
    
        Ok(self.clone())
    }
    


    fn update(self: &Rc<Self>, reqdata: Value, ctrl: Value) -> Result<Rc<Self>, SolardemoError> {
        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some("update".to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqdata: Some(reqdata),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );
    
        self.run_op(&ctx, &|ctx| {
            if let Some(result) = ctx.result.borrow().clone() {
                let (resmatch, resdata) = {
                    let r = result.borrow();
                    (r.resmatch.clone(), r.resdata.clone())
                };
                if let Value::Map(_) = resmatch {
                    *self.mtch.borrow_mut() = resmatch;
                }
                if !resdata.is_noval() && !resdata.is_null() {
                    *self.data.borrow_mut() = match to_map(&vs::clone(&resdata)) {
                        Value::Map(m) => Value::Map(m),
                        _ => Value::empty_map(),
                    };
                }
            }
        })?;
    
        // The operation resolves to THIS entity: `run_op` has just absorbed the
        // result into it, and the caller reaches the record through `.data(None)`.
        // See AGENTS.md "Entity operations return ENTITIES".
    
        Ok(self.clone())
    }
    


    fn remove(self: &Rc<Self>, reqmatch: Value, ctrl: Value) -> Result<Rc<Self>, SolardemoError> {
        let ctx = self.utility.make_context(
            CtxSpec {
                opname: Some("remove".to_string()),
                ctrl: Some(ctrl),
                mtch: Some(self.mtch.borrow().clone()),
                data: Some(self.data.borrow().clone()),
                reqmatch: Some(reqmatch),
                ..Default::default()
            },
            Some(&self.ent_ctx()),
        );
    
        self.run_op(&ctx, &|ctx| {
            if let Some(result) = ctx.result.borrow().clone() {
                let (resmatch, resdata) = {
                    let r = result.borrow();
                    (r.resmatch.clone(), r.resdata.clone())
                };
                if let Value::Map(_) = resmatch {
                    *self.mtch.borrow_mut() = resmatch;
                }
                if !resdata.is_noval() && !resdata.is_null() {
                    *self.data.borrow_mut() = match to_map(&vs::clone(&resdata)) {
                        Value::Map(m) => Value::Map(m),
                        _ => Value::empty_map(),
                    };
                }
            }
        })?;
    
        // The operation resolves to THIS entity: `run_op` has just absorbed the
        // result into it, and the caller reaches the record through `.data(None)`.
        // See AGENTS.md "Entity operations return ENTITIES". A removed entity
        // keeps its data but is no longer a live record.
        self.mark_deleted();
    
        Ok(self.clone())
    }
    
}
